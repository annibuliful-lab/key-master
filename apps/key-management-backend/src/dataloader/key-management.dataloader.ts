import { prismaClient } from '@key-master/db';
import { mapDataWithIdsByCustomFieldId } from '@key-master/graphql';
import DataLoader from 'dataloader';

export const keyManagementDataLoader = new DataLoader(async (ids: string[]) => {
  const keyManagements = await prismaClient.keyManagment.findMany({
    where: { id: { in: [...new Set(ids)] } },
  });

  return mapDataWithIdsByCustomFieldId({
    data: keyManagements,
    idField: 'id',
    ids,
  });
});
