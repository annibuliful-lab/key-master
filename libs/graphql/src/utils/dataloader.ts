export function mapDataWithIds<T extends { id: string }>(
  data: T[],
  ids: string[]
) {
  const dataMap = new Map(data.map((d) => [d.id, d]));
  return ids.map((id) => dataMap.get(id));
}

interface IMapDataWithIdsByCustomFieldId<T> {
  data: T[];
  ids: string[];
  idField: keyof T;
}

export function mapDataWithIdsByCustomFieldId<T>({
  data,
  idField,
  ids,
}: IMapDataWithIdsByCustomFieldId<T>) {
  const dataMap = new Map(
    data.map((d) => [d[idField] as unknown as string, d])
  );
  return ids.map((id) => dataMap.get(id));
}
