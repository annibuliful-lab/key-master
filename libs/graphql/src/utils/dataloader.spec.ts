import { mapDataWithIds, mapDataWithIdsByCustomFieldId } from './dataloader';

describe('dataloader', () => {
  describe('mapDataWithIds', () => {
    it('returns map data structure', () => {
      expect(
        mapDataWithIds([{ id: '1' }, { id: '2' }], ['1', '2'])
      ).toMatchSnapshot();
    });
  });

  describe('mapDataWithIdsByCustomFieldId', () => {
    it('returns map data structure with custom field', () => {
      expect(
        mapDataWithIdsByCustomFieldId({
          data: [
            { id: '1', name: 'AAAA' },
            { id: '2', name: 'BBBB' },
            { id: 'a', name: 'CCCC' },
          ],
          ids: ['1', '2'],
          idField: 'id',
        })
      ).toMatchSnapshot();
    });
  });
});
