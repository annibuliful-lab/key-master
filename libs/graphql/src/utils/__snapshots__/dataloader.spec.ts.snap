// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataloader mapDataWithIds returns map data structure 1`] = `
Array [
  Object {
    "id": "1",
  },
  Object {
    "id": "2",
  },
]
`;

exports[`dataloader mapDataWithIdsByCustomFieldId returns map data structure with custom field 1`] = `
Array [
  Object {
    "id": "1",
    "name": "AAAA",
  },
  Object {
    "id": "2",
    "name": "BBBB",
  },
]
`;
